src/microbe_stage/PlacedOrganelle.cs

```diff
@@ -328,14 +328,30 @@ public Transform3D CalculateVisualsTransform()
 
     public Transform3D CalculateVisualsTransformExternal(Vector3 externalPosition, Quaternion orientation)
     {
+        float scaleZ = 0.0f;
+
+        if (Upgrades?.CustomUpgradeData is FlagellumUpgrades flagellumUpgrades)
+        {
+            var flagellumLength = flagellumUpgrades.LengthFraction;
+
+            scaleZ = Constants.FLAGELLA_MAX_UPGRADE_VISUAL_LENGTH * flagellumLength
+                + Constants.FLAGELLA_MIN_VISUAL_LENGTH;
+        }
+
         var scale = CalculateTransformScale();
 
         cachedExternalOrientation = orientation;
         cachedExternalPosition = externalPosition;
 
         // TODO: check that the rotation of ModelOffset works correctly here (also in
         // CalculateVisualsTransformExternalCached)
-        return new Transform3D(new Basis(orientation).Scaled(new Vector3(scale, scale, scale)),
+        if (scaleZ == 0.0f)
```

Comment:
Couldn't scaleZ start off by being equal to scale? That way this extra if and duplicated complicated math line is avoided.